dlt/destinations/snowflake/snowflake.py

```diff
@@ -48,44 +51,76 @@
     "VARIANT": "complex"
 }
 
+
 class SnowflakeLoadJob(LoadJob, FollowupJob):
     def __init__(
             self, file_path: str, table_name: str, write_disposition: TWriteDisposition, load_id: str, client: SnowflakeSqlClient,
-            stage_name: Optional[str] = None, keep_staged_files: bool = True
+            stage_name: Optional[str] = None, keep_staged_files: bool = True, forward_staging_credentials: bool = True, staging_config: Optional[DestinationClientStagingConfiguration] = None
     ) -> None:
         file_name = FileStorage.get_file_name_from_file_path(file_path)
         super().__init__(file_name)
 
         with client.with_staging_dataset(write_disposition == "merge"):
             qualified_table_name = client.make_qualified_table_name(table_name)
 
-            if stage_name:
+            # extract and prepare some vars
+            bucket_path = CopyFileLoadJob.get_bucket_path(file_path) if NewReferenceJob.is_reference_job(file_path) else ""
+            file_name = FileStorage.get_file_name_from_file_path(bucket_path) if bucket_path else file_name
+            from_clause = ""
+            credentials_clause = ""
+            files_clause = ""
+            stage_file_path = ""
+
+            # create storage integration stage for reference jobs if defined, for now gcs only
+            if bucket_path and stage_name:
+                from_clause = f"FROM @{stage_name}/"
+                # select all files that match the file from our job
+                bucket = CopyFileLoadJob.get_bucket(file_path)
+                file_path = bucket_path.replace(bucket + "/", "")
+                files_clause = f"FILES = ('{file_path}')"
+            # s3 credentials case
+            elif bucket_path:
+                if bucket_path.startswith("s3://") and forward_staging_credentials:
+                    aws_access_key = staging_config.credentials.aws_access_key_id # type: ignore
+                    aws_secret_key = staging_config.credentials.aws_secret_access_key # type: ignore
+                    credentials_clause = f"""CREDENTIALS=(AWS_KEY_ID='{aws_access_key}' AWS_SECRET_KEY='{aws_secret_key}')"""
+                from_clause = f"FROM '{bucket_path}'"
+            # create a stage if so defined
+            elif stage_name:
```

Comment:
I'd always assume that stage exists if name is provided. so we can drop this part and simplify if condition

Reply:
I am not quite sure why this is done this way, this was like this before I started working on snowflake, so I don't understand the reasoning tbh. I will keep it for now and if you insist I will change it :)

Reply:
heh when we wrote the code I didn't know exact how stages work in snowflake. there's no use case where creating a named stage makes sense.
- if stage_name is set we assume that it exist and use it
- if it is not set and we have a bucket url (external files) then we fail the load (terminal exception) with a user friendly message
- for local load and no stage we use built in table stage (we do it now)